### PR TITLE
fix(time_offset): improved LIMIT-handling in advanced analytics

### PR DESCRIPTION
**Pull request by Antonio-RiveroMartnez**
### SUMMARY
When applying a time_offset in Advanced Analytics, if the row_limit is too small, data will not be present when the join occurs, so this PR prevent such behavior by not using limit nor offset in the inner query (the shifted one).

For example this query shown in the following video has all the records for a period of 10+ years:

https://github.com/apache/superset/assets/38889534/8dca8b16-5c05-4b2d-8fd6-67387c8708e7

We can see that the result of shifted queries not always include the same periods than the original query, so when the join gets applied, it won't match some records, thus giving you incorrect output.

Original query with limit and sorting:
Captura de pantalla 08 04 2024 a 11 05 13 AM

1 year ago query with limit and sorting:
Captura de pantalla 08 04 2024 a 11 05 50 AM

2 years ago query with limit and sorting:
Captura de pantalla 08 04 2024 a 11 06 08 AM

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Error:
In the following image some datapoints are missing in the shifted columns because they are not part of the result dataframe after sorting and limiting it like the original query.
error

Fix:
Now we can see that the query is able to find the records that belong to those offsets regardless, because it's joining using the complete data.
test

### ADDITIONAL INFORMATION

- [X] Has associated issue: https://github.com/apache/superset/issues/27504
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/common/query_context_processor.py [hunk 1]
```diff
@@ -455,6 +455,13 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
                 for metric in metric_names
             }
 
+            # When the original query has limit or offset we wont apply those
+            # to the subquery so we prevent data inconsistency due to missing records
+            # in the dataframes when performing the join
+            if query_object.row_limit or query_object.row_offset:
+                query_object_clone_dct["row_limit"] = None
```
Comment (zhaoyongjie) [inline superset/common/query_context_processor.py:462]: We should apply `row_limit` in config.py, in order to avoid a heavy query.

Comment (Antonio-RiveroMartnez) [reply superset/common/query_context_processor.py:462]: makes sense, will do. Thanks

**Comment by john-bodley** [comment]
@Antonio-RiveroMartnez are there any open issues we can link to this? I'm somewhat surprised that no one has reported an issue with this in the past.

**Comment by Antonio-RiveroMartnez** [comment]
Hey @john-bodley I just linked it in the PR description, thanks.

File: superset/common/query_context_processor.py [hunk 2]
```diff
@@ -455,6 +455,13 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
                 for metric in metric_names
             }
 
+            # When the original query has limit or offset we wont apply those
```
Comment (mistercrunch) [inline superset/common/query_context_processor.py:458]: Wow this method has gotten large! Thinking about refactors here and factoring out a `processing_time_offset` (a method for each time offset) might help a little in reading this, but probably breaking down further would be good too. 

Doing the refactor in this PR as part of this PR would make it harder to spot the diff, so I'll approve this one, but hoping we can refactor some of these into more bite-sized methods. Maybe a goal could be to remove the `disable=too-many-locals,too-many-statements` above.

**Review by mistercrunch — APPROVED** [review]
Changes + tests LGTM. Let's follow up with a PR breaking this large method in the right places.

**Review by eschutho — APPROVED** [review]

**Pull request merged**